Avoid importing executor during conf validation

uranusjr opened the pull request:
Loading an executor is heavy work that is only needed in a scheduler. We should avoid doing that.

This moves the validation to when the executor is actually loaded; this slightly degrades the UX when there's an invalid configuration (the scheduler fails later), but the gain is worthwhile IMO. For the webserver, for example, the startup time of a "bare" Airflow installation (no providers or plugins) is down >25% on my machine (roughly 8s to 6s).

uranusjr committed:
Split executor and database conf validation

We're going to skip some of the checks, and this makes the effort
slightly easier.

uranusjr committed:
Avoid importing executor during conf validation

Loading an executor is heavy work that is only needed in a scheduler.
We should avoid doing that.

This moves the validation to when the executor is actually loaded; this
slightly degrades the UX when there's an invalid configuration (the
scheduler fails later), but the gain is worthwhile IMO. For the
webserver, for example, the startup time of a "bare" Airflow installation
(no providers or plugins) is down >25% on my machine (roughly 8s to 6s).

andriisoldatenko commented on airflow/executors/executor_loader.py:
```diff
@@ -41,6 +41,22 @@
     from airflow.executors.base_executor import BaseExecutor
 
 
+def _validate_database_executor_compatibility(executor: type[BaseExecutor]) -> None:
```
do we want to write tests for this function?

uranusjr replied:
Added

ashb commented on airflow/configuration.py:
```diff
@@ -429,30 +428,34 @@ def _validate_enums(self):
         """
-        executor, _ = ExecutorLoader.import_default_executor_cls()
-        is_sqlite = "sqlite" in self.get("database", "sql_alchemy_conn")
+        self._validate_sqlite3_version()
```
Rather than calling this from in here, call this function directly from The existing `validate()` function.

ashb reviewed (commented):
With a bit more refactoring we could validate this config inside the scheduler CLI command before deamonizing.

Changes that would be needed for that:

- In `airflow/cli/commands/scheduler_command.py` create the SchedulerJob in `scheduler` function (instead of in `_run_scheduler_job`)
- Call `scheduler_job.executor` to force the load/vivify the cached property, that'll die if it's not valid.

uranusjr committed:
Remove redundant abstraction

uranusjr committed:
Refactor validation structure

The validation logic is now moved into the loader as a public function,
with tests. Validation is still done when an executor is loaded, but we
also call this when a SchedulerJob is instantiated eagerly. This is
strictly not necessary, but makes the code more readable. The extra
validation call is also very cheap because we cache the validation
result.

uranusjr committed:
Fix /run test for CI

I don't know how this ever worked, but the view is totally broken.

uranusjr commented on tests/www/views/test_views_tasks.py:
```diff
@@ -492,7 +496,7 @@ def test_code_from_db_all_example_dags(admin_client):
                 ignore_all_deps="false",
                 ignore_ti_state="true",
-                execution_date=DEFAULT_DATE,
+                dag_run_id=DEFAULT_DAGRUN,
```
How did this ever work before? The view does not accept `execution_date` (not anymore, I think) but only `dag_run_id`.

Taragolis replied:
I guess we use mock client rather than actually call views here: https://flask.palletsprojects.com/en/2.2.x/testing/#sending-requests-with-the-test-client

https://github.com/apache/airflow/blob/f17e2ba48b59525655a92e04684db664a672918f/tests/www/views/conftest.py#L64-L65

uranusjr replied:
I think the mock client still executes the views (it simply skips the HTTP and directly calls the routers).

uranusjr committed:
Mock out executor in sensor tests

uranusjr committed:
Implement global switch to disable check in tests
